compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs

```diff
@@ -247,31 +257,36 @@ generateTemplateInstance env typeCon typeParams externPkgId =
                             noLoc $
                             HsAppTy noExt templateTy $
                             noLoc $
-                            convType env $ lfTemplateType typeCon typeParams
+                            convType env lfTemplateType
                       }
-            , cid_binds = listToBag $ map (classMethodStub typeCon) templateMethodNames
+            , cid_binds = listToBag $ map classMethodStub methodMapping
             , cid_sigs = []
             , cid_tyfam_insts = []
             , cid_datafam_insts = []
             , cid_overlap_mode = Nothing
             }
   where
     moduleNameStr = T.unpack $ LF.moduleNameString $ LF.moduleName $ envMod env
-    moduleName0 =
-        LF.ModuleName $
-        map T.pack $
-        splitOn "." moduleNameStr
+    moduleName0 = LF.moduleName $ envMod env
     templateTy =
         noLoc $
         HsTyVar noExt NotPromoted $
         noLoc $
         mkRdrQual (mkModuleName "DA.Internal.Template") $
         mkOccName varName "Template" :: LHsType GhcPs
-    lfTemplateType dataTypeCon dataParams =
+    lfTemplateType =
         LF.mkTApps
-            (LF.TCon (LF.Qualified LF.PRSelf moduleName0 dataTypeCon))
-            (map (LF.TVar . fst) dataParams)
-    templateMethodNames =
+            (LF.TCon (LF.Qualified LF.PRSelf moduleName0 typeCon))
+            (map (LF.TVar . fst) typeParams)
+    methodMapping =
+        map (\funName -> (funName, mkExternalString funName)) methodNames
+    mkExternalString :: T.Text -> String
+    mkExternalString funName =
```

Comment:
we have copies of `lfTemplateType`, `mkExternalString` and `methodMapping` in `generateChoiceInstance` and `generateTemplateInstance`. We should factor these out.

Reply:
done, except for `mkExternalString` which varies